feature-odyssey/src/main/kotlin/com/hedvig/android/odyssey/step/singleitem/SingleItemViewModel.kt

```diff
@@ -0,0 +1,367 @@
+import com.hedvig.android.odyssey.data.ClaimFlowRepository
+import com.hedvig.android.odyssey.data.ClaimFlowStep
+import com.hedvig.android.odyssey.navigation.ClaimFlowDestination
+import com.hedvig.android.odyssey.navigation.ItemBrand
+import com.hedvig.android.odyssey.navigation.ItemModel
+import com.hedvig.android.odyssey.navigation.ItemProblem
+import com.hedvig.android.odyssey.navigation.UiMoney
+import com.hedvig.android.odyssey.ui.DatePickerUiState
+import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.flow.SharingStarted
+import kotlinx.coroutines.flow.StateFlow
+import kotlinx.coroutines.flow.WhileSubscribed
+import kotlinx.coroutines.flow.combine
+import kotlinx.coroutines.flow.stateIn
+import kotlinx.coroutines.flow.update
+import kotlinx.coroutines.launch
+import kotlinx.datetime.Clock
+import kotlinx.datetime.Instant
+import kotlinx.datetime.TimeZone
+import kotlinx.datetime.toLocalDateTime
+import octopus.type.CurrencyCode
+import octopus.type.FlowClaimItemBrandInput
+import octopus.type.FlowClaimItemModelInput
+import kotlin.time.Duration.Companion.seconds
+
+internal class SingleItemViewModel(
+  singleItem: ClaimFlowDestination.SingleItem,
+  private val claimFlowRepository: ClaimFlowRepository,
+  clock: Clock = Clock.System,
+) : ViewModel() {
+
+  private val itemBrandsUiState: MutableStateFlow<ItemBrandsUiState> =
+    MutableStateFlow(ItemBrandsUiState.fromSingleItem(singleItem))
+  private val itemModelsUiState: MutableStateFlow<ItemModelsUiState> =
+    MutableStateFlow(ItemModelsUiState.fromSingleItem(singleItem))
+
+  // Holds most of what [uiState] does, minus what's inside brands and models UiState above
+  private val partialUiState: MutableStateFlow<PartialSingleItemUiState> =
+    MutableStateFlow(PartialSingleItemUiState.fromSingleItem(singleItem, clock))
+
+  val uiState: StateFlow<SingleItemUiState> = combine(
+    itemBrandsUiState,
+    itemModelsUiState,
+    ::transformItemModelsToOnlyContainModelsOfTheSelectedBrand,
+  ).combine(partialUiState) { (itemBrandsUiState, itemModelsUiState), partialUiState ->
+    SingleItemUiState(
+      partialUiState.datePickerUiState,
+      partialUiState.purchasePriceUiState,
+      itemBrandsUiState,
+      itemModelsUiState,
+      partialUiState.itemProblemsUiState,
+      partialUiState.isLoading,
+      partialUiState.hasError,
+      partialUiState.nextStep,
+    )
+  }.stateIn(
+    viewModelScope,
+    SharingStarted.WhileSubscribed(5.seconds),
+    SingleItemUiState.fromInitialSingleItem(singleItem, clock),
+  )
+
+  private fun transformItemModelsToOnlyContainModelsOfTheSelectedBrand(
+    itemBrandsUiState: ItemBrandsUiState,
+    itemModelsUiState: ItemModelsUiState,
+  ): Pair<ItemBrandsUiState, ItemModelsUiState> {
+    if (itemModelsUiState !is ItemModelsUiState.Content) return itemBrandsUiState to itemModelsUiState
+    val selectedBrandId = itemBrandsUiState
+      .asContent()
+      ?.selectedItemBrand
```

Comment:
Without the ID field in the interface, this check would need to look instead of: 
```kotlin
val selectedBrandId = itemBrandsUiState
      .asContent()
      ?.selectedItemBrand
      ?.asKnown()
      ?.itemBrandId
      ?: return itemBrandsUiState to itemModelsUiState
```
it'd look like this:
```kotlin
val selectedItemBrand: ItemBrand = itemBrandsUiState
  .asContent()
  ?.selectedItemBrand ?: return itemBrandsUiState to itemModelsUiState
val selectedBrandId = when(selectedItemBrand) {
  ItemBrand.Unknown -> return itemBrandsUiState to itemModelsUiState
  is ItemBrand.Known -> selectedItemBrand.itemBrandId
}
```
Since we also need to check if the itemBrandsUiState has content (it doesn't when backend returns an empty list of brands), along with doing this check about if the selected item is not the "Not sure" item (hence the asKnown() check)

So all in all I do kinda like that we're doing all the checks from start to finish, and proceed with the nullable type so that if all of them succeed and get get not-null back we're sure it's selected, otherwise no matter which one fails we exit quick by not filtering the items at all.

Now for the ID in the interface, that'd make this look like:
```kotlin
val selectedBrandId = itemBrandsUiState
      .asContent()
      ?.selectedItemBrand
      ?.itemBrandId
      ?: return itemBrandsUiState to itemModelsUiState
```
So we're saving 1 `asKnown()` check but making the interface bigger, I don't think this tradeoff is worth it.

All in all, to get the `selectedItemBrand` as in your snippet, we simply must check if `ItemBrandsUiState` is in the "Content" state, so we will have to have a bunch of checks no matter what we do, and making them all in 1 chained check feels quite okay to me tbh 👀 I don't think there's too much room to simplify this without complicating some other part, wydt?

Reply:
This is how that PR with the ID would look like btw, not too much to gain as I see it 👀
<img width="1029" alt="image" src="https://user-images.githubusercontent.com/44558292/227264526-d736bed2-d0c3-4664-a510-94094cc3738c.png">
<img width="1030" alt="image" src="https://user-images.githubusercontent.com/44558292/227264560-71db2a40-a431-4dad-bfcb-37a24008ee9a.png">